base/applications/mspaint/textedit.cpp

```diff
@@ -10,50 +10,477 @@
 
 #include "precomp.h"
 
+#define CXY_GRIP 4
+
 /* FUNCTIONS ********************************************************/
-LRESULT CTextEditWindow::OnCreate(UINT nMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+
+SIZE CTextEditWindow::DoCalcRect(HDC hDC, LPTSTR pszText, INT cchText,
+                                 LPRECT prcParent, LPCTSTR pszOldText)
+{
+    RECT rc;
+    GetClientRect(&rc);
+    MapWindowPoints(NULL, (LPPOINT)&rc, 2);
+
+    TEXTMETRIC tm;
+    GetTextMetrics(hDC, &tm);
+
+    DWORD dwMargin = (DWORD)DefWindowProc(EM_GETMARGINS, 0, 0);
+    LONG leftMargin = LOWORD(dwMargin), rightMargin = HIWORD(dwMargin);
+
+    // Calculate width and height of the text
+    INT x = leftMargin, y = tm.tmHeight, xMax = x, yMax = y;
+    INT ich;
+    for (ich = 0; ich < cchText; ++ich)
+    {
+        if (pszText[ich] == TEXT('\r'))
+            continue;
+
+        if (pszText[ich] == TEXT('\n'))
+        {
+            if (rc.top + yMax + tm.tmHeight > prcParent->bottom)
+            {
+                pszText[ich] = 0; // Truncate
+                break;
+            }
+            x = leftMargin;
+            y += tm.tmHeight;
+            if (yMax < y)
+                yMax = y;
+            continue;
+        }
+
+        SIZE siz;
+        GetTextExtentPoint32(hDC, &pszText[ich], 1, &siz);
+
+        // TODO: Tabs
+        x += siz.cx;
+        if (xMax < x)
+            xMax = x;
+    }
+
+    if (ich != cchText && pszOldText) // Truncated
+    {
+        DWORD iStart, iEnd;
+        SendMessage(EM_GETSEL, (WPARAM)&iStart, (LPARAM)&iEnd);
+        SetWindowText(pszOldText);
+        SendMessage(EM_SETSEL, iStart, iEnd);
+        MessageBeep(0xFFFFFFFF);
+        cchText = ich;
+    }
+
+    // Consider italic overhang
+    ABCFLOAT WidthsABC;
+    FLOAT overhang = 0;
+    if (cchText > 0)
+    {
+        GetCharABCWidthsFloat(hDC, pszText[cchText - 1], pszText[cchText - 1], &WidthsABC);
+        overhang = WidthsABC.abcfC;
+        if (overhang > 0)
+            overhang = 0;
+    }
+
+    SIZE ret = { xMax + rightMargin - LONG(overhang) + 1, yMax + tm.tmDescent };
+    return ret;
+}
+
+#define X0 rc.left
+#define X1 ((rc.left + rc.right - CXY_GRIP) / 2)
+#define X2 (rc.right - CXY_GRIP)
+#define Y0 rc.top
+#define Y1 ((rc.top + rc.bottom - CXY_GRIP) / 2)
+#define Y2 (rc.bottom - CXY_GRIP)
+#define RECT0 X0, Y0, X0 + CXY_GRIP, Y0 + CXY_GRIP
+#define RECT1 X1, Y0, X1 + CXY_GRIP, Y0 + CXY_GRIP
+#define RECT2 X2, Y0, X2 + CXY_GRIP, Y0 + CXY_GRIP
+#define RECT3 X0, Y1, X0 + CXY_GRIP, Y1 + CXY_GRIP
+#define RECT4 X2, Y1, X2 + CXY_GRIP, Y1 + CXY_GRIP
+#define RECT5 X0, Y2, X0 + CXY_GRIP, Y2 + CXY_GRIP
+#define RECT6 X1, Y2, X1 + CXY_GRIP, Y2 + CXY_GRIP
+#define RECT7 X2, Y2, X2 + CXY_GRIP, Y2 + CXY_GRIP
```

Comment:
The RECT macros could use some explanation. It took me a moment to realize that they expand to a partial parameter list.

Reply:
Comments added.